consensus/hotstuff/signature.go

```diff
@@ -48,50 +48,49 @@ func (t SigType) Valid() bool {
 	return t == SigTypeStaking || t == SigTypeRandomBeacon
 }
 
-// WeightedSignatureAggregator aggregates signatures of the same signature scheme from different signers.
-// The module is aware of weights assigned to each signer.
-// It can either be used to aggregate staking signatures or aggregate random beacon signatures,
-// but not a mix of both.
+// WeightedSignatureAggregator aggregates signatures of the same signature scheme and the same message from different signers.
+// The public keys and message are aggreed upon upfront.
+// It is also recommended to only aggregate signatures generated with keys representing equivalent security-bit level.
+// The module is aware of weights assigned to each signer, as well as a total weight threshold.
 // Implementation of SignatureAggregator must be concurrent safe.
 type WeightedSignatureAggregator interface {
 	// Verify verifies the signature under the stored public key corresponding to the signerID, and the stored message.
 	Verify(signerID flow.Identifier, sig crypto.Signature) (bool, error)
 
-	// TrustedAdd adds an already verified signature, with weight for the given signer,
-	// and add it to the total weight, and returns the total weight that have been collected.
-	// return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
-	//   (1000 is just an example)
-	// return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
+	// TrustedAdd adds a signature to the internal set of signatures.
+	// It adds the signer's weight to the total collected weight and returns the total weight regardless
+	// of the returned error.
```

Comment:
To me, the comment was a bit confusing depending on precedence of the `and` (which is not clearly defined in plain english). You could read it like 
* (It adds the signer's weight to the total collected weight and returns the total weight)
   
   regardless of the returned error.

```suggestion
	// TrustedAdd adds a signature to the internal set of signatures and adds the signer's weight to the total collected weight, iff the signature is _not_ a duplicate. 
	// The total weight of all collected signatures (excluding duplicates) is returned regardless
	// of any returned error.
```

### Suggestion:
Specify a sentinel error for duplicated signatures. Depending on the usage patter of the higher-level business logic, a duplicated signature could be a fatal bug or expected.

Reply:
BTW, in my mind 
1. signerID  corresponds to a known consensus participant but the signature is cryptographically invalid
2. signerID  does not correspond to a known consensus participant
3. duplicated signature

could _all_  be _expected_ during normal operations (depending on the usage patter of the higher-level business logic). 

I would suggest to follow the same API convention in all three cases, for all methods. Therefore, I would suggest to use the following convention:
* represent each case by a dedicated sentinel error 
* `TrustedAdd` can return any of the 3 sentinels 
* Change `Verify` to 
   ```golang 
   Verify(signerID flow.Identifier, sig crypto.Signature) error
   ```
    * it only returns nil if the signature is valid
    * case 1. and 2. are represented by sentinel errors
    * case 3 is irrelevant for this method 




I would suggest to update the API